src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -1,243 +1,249 @@
 open Ppxlib
+module G = Qcheck_generators
+module O = G.Observable
 
-(** TypeGen can serve as a derivation environment. The map can be used
-    to remember how a type should be translated.
+(** {1. ppx_deriving_qcheck} *)
 
-    For instance, a recursive type must be derivated to a self recursive
-    call reducing the size.
+(** ppx_deriving_qcheck is a ppx deriver for QCheck generators. It does a traversal
+    map on type declarations annoted with [QCheck].
 
+    Example:
     {[
-    type tree = Leaf of int | Node of tree * tree
+    module Tree : sig
+      type t
+
+      val gen : t QCheck.Gen.t
+    end = struct
+      type t = Leaf | Node of int * tree * tree
+      [@@deriving qcheck]
+    end
     ]}
+*)
+
+(** {2. Misc. helpers} *)
+
+(** [name s] produces the generator name based on [s] *)
+let name ?(sized = false) s =
+  let prefix = "gen" in
+  (match s with "t" -> prefix | s -> prefix ^ "_" ^ s) ^
+    (if sized then "_sized" else "")
+
+(** [pat ~loc s] creates a pattern for a generator based on {!name}. *)
+let pat ~loc ?sized s =
+  let (module A) = Ast_builder.make loc in
+  let s = name ?sized s in
+  A.pvar s
 
-    becomes
+(** {2. Recursive generators} *)
+
+(** Recursive generators must be treated separatly:
 
     {[
-    let gen_tree =
-      let open QCheck in
-      let open Gen in
-      sized
-        @@ fix (fun self -> function
-             | 0 -> frequency [ (1, pure Leaf) ]
-             | n ->
-                 frequency
-                   [
-                     (1, pure Leaf);
-                     ( 1,
-                       map
-                         (fun (gen0, gen1, gen2) -> Node (gen0, gen1, gen2))
-                         (triple int (self (n / 1)) (self (n / 1))) );])
+    type 'a list = Cons of 'a * 'a list | Nil 
     ]}
 
-    The type [tree] is stored in a TypeGen.t with tree <- [%expr self (n/2)]. This
-    avoids the case where [tree] is derivated to [gen_tree]
-*)
-module TypeGen = Map.Make (struct
-  type t = string
+    becomes:
+
+    {[
+    let rec gen_list_sized gen_a n =
+      match n with
+      | 0 -> pure Nil
+      | n -> map2 (fun x xs -> Cons (x, xs) gen_a (gen_list_sized gen_a (n/2))
+
+    let gen_list_sized gen_a = sized @@ (gen_list_sized gen_a)
+    ]}
 
-  let compare = compare
-end)
+    In the basic derivation {[ 'a list ]} would be translated to {[gen_list]}. However,
+    we want the generator to call itsef.
+ *)
 
 let rec longident_to_str = function
   | Lident s -> s
   | Ldot (lg, s) -> Printf.sprintf "%s.%s" (longident_to_str lg) s
   | Lapply (lg1, lg2) ->
       Printf.sprintf "%s %s" (longident_to_str lg1) (longident_to_str lg2)
 
-let rec is_rec_typ typ_name = function
+(** [is_rec_type_decl typ_names typ] looks for elements of [typ_names]
+    recursively in [typ]. *)
+let rec is_rec_type_decl typ_names typ =
```

Comment:
very minor:
these 3 mutually recursive functions could reordered
```ocaml
let rec is_rec_type ...
let is_rec_constr_decl ,,,
let is_rec_type_decl ...
```
so that only the first is recursive, no?

Reply:
You are right, 4eca3a41b78fdc49ea1867291afade5c75960624.